src/realm/parser/query_builder.cpp

```diff
@@ -37,131 +37,129 @@ using namespace query_builder;
 }
 
-template<typename T>
-void do_add_null_comparison_to_query(Query &query, Predicate::Operator op, const PropertyExpression &expr)
+template <typename T>
+Query do_make_null_comparison_query(Predicate::Operator op, const PropertyExpression& expr)
 {
     Columns<T> column = expr.link_chain_getter().template column<T>(expr.get_dest_col_key());
     switch (op) {
         case Predicate::Operator::NotEqual:
-            query.and_query(column != realm::null());
-            break;
+            return column != realm::null();
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
         case Predicate::Operator::Equal:
-            query.and_query(column == realm::null());
-            break;
+            return column == realm::null();
         default:
-            throw std::logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
+            throw_logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
     }
 }
 
 template <typename T>
-void do_add_null_comparison_to_query(Query& query, Predicate::Operator op, const PrimitiveListExpression& expr)
+Query do_make_null_comparison_query(Predicate::Operator op, const PrimitiveListExpression& expr)
 {
     Columns<Lst<T>> column = expr.value_of_type_for_query<T>();
     switch (op) {
         case Predicate::Operator::NotEqual:
-            query.and_query(column != realm::null());
-            break;
+            return column != realm::null();
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
         case Predicate::Operator::Equal:
-            query.and_query(column == realm::null());
-            break;
+            return column == realm::null();
         default:
-            throw std::logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
+            throw_logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
     }
 }
 
 template <>
-void do_add_null_comparison_to_query<Link>(Query&, Predicate::Operator, const PrimitiveListExpression&)
+REALM_FORCEINLINE Query do_make_null_comparison_query<Link>(Predicate::Operator, const PrimitiveListExpression&)
 {
-    throw std::logic_error("Invalid query, list of primitive links is not a valid Realm contruct");
+    throw_logic_error("Invalid query, list of primitive links is not a valid Realm contruct");
 }
 
-template<>
-void do_add_null_comparison_to_query<Link>(Query &query, Predicate::Operator op, const PropertyExpression &expr)
+template <>
+Query do_make_null_comparison_query<Link>(Predicate::Operator op, const PropertyExpression& expr)
 {
     REALM_ASSERT_DEBUG(!expr.dest_type_is_backlink()); // this case should be handled at a higher layer
     switch (op) {
-        case Predicate::Operator::NotEqual:
-            query.Not();
-            REALM_FALLTHROUGH;
+        case Predicate::Operator::NotEqual: {
```

Comment:
This is certainly a win in terms of readability. It might have been intending to use a potentially more optimized path for equality queries and inverting the result via "Not()" rather than using the NotEqual operator. If it doesn't show up in the performance report then it isn't something we have explicitly optimized for and I'd be fine with the change. The problem is that it appears the performance report is not displaying the results of this PR due to a bug in the performance scripts. I'll see if I can fix that so we can view the results.

Reply:
The performance report [here](https://ci.realm.io/job/realm/job/realm-core/job/PR-3858/3/Performance_5fReport/) from https://github.com/realm/realm-core/pull/3858 indicates that we are either not explicitly tracking this code path, or it is has negligible performance impact so I'm satisfied with that.